collectors/ebpf.plugin/ebpf_socket.c

```diff
@@ -0,0 +1,381 @@
+    }
+}
+
+/*****************************************************************
+ *
+ *  FUNCTIONS TO CREATE CHARTS
+ *
+ *****************************************************************/
+
+/**
+ * Create global charts
+ *
+ * Call ebpf_create_chart to create the charts for the collector.
+ *
+ * @param em a pointer to the structure with the default values.
+ */
+static void ebpf_create_global_charts(ebpf_module_t *em) {
+    ebpf_create_chart(NETDATA_EBPF_FAMILY
+        , NETDATA_TCP_FUNCTION_COUNT
+        , "Calls"
+        , NETDATA_SOCKET_GROUP
+        , 950
+        , ebpf_create_global_dimension
+        , socket_publish_aggregated
+        , 3);
+
+    ebpf_create_chart(NETDATA_EBPF_FAMILY
+        , NETDATA_TCP_FUNCTION_BYTES
+        , "bytes/s"
+        , NETDATA_SOCKET_GROUP
+        , 951
+        , ebpf_create_global_dimension
+        , socket_publish_aggregated
+        , 3);
+
+    if (em->mode < MODE_ENTRY) {
+        ebpf_create_chart(NETDATA_EBPF_FAMILY
+            , NETDATA_TCP_FUNCTION_ERROR
+            , "Calls"
+            , NETDATA_SOCKET_GROUP
+            , 952
+            , ebpf_create_global_dimension
+            , socket_publish_aggregated
+            , 2);
+    }
+
+    ebpf_create_chart(NETDATA_EBPF_FAMILY
+        , NETDATA_UDP_FUNCTION_COUNT
+        , "Calls"
+        , NETDATA_SOCKET_GROUP
+        , 953
+        , ebpf_create_global_dimension
+        , &socket_publish_aggregated[NETDATA_UDP_START]
+        , 2);
+
+    ebpf_create_chart(NETDATA_EBPF_FAMILY
+        , NETDATA_UDP_FUNCTION_BYTES
+        , "bytes/s"
+        , NETDATA_SOCKET_GROUP
+        , 954
+        , ebpf_create_global_dimension
+        , &socket_publish_aggregated[NETDATA_UDP_START]
+        , 2);
+
+    if (em->mode < MODE_ENTRY) {
+        ebpf_create_chart(NETDATA_EBPF_FAMILY
+            , NETDATA_UDP_FUNCTION_ERROR
+            , "Calls"
+            , NETDATA_SOCKET_GROUP
+            , 955
+            , ebpf_create_global_dimension
+            , &socket_publish_aggregated[NETDATA_UDP_START]
+            , 2);
+    }
+}
+
+/*****************************************************************
+ *
+ *  READ INFORMATION FROM KERNEL RING
+ *
+ *****************************************************************/
+
+/**
+ * Read the hash table and store data to allocated vectors.
+ */
+static void read_hash_global_tables()
+{
+    uint64_t idx;
+    netdata_idx_t res[NETDATA_SOCKET_COUNTER];
+
+    netdata_idx_t *val = socket_hash_values;
+    for (idx = 0; idx < NETDATA_SOCKET_COUNTER ; idx++) {
+        if (!bpf_map_lookup_elem(map_fd[4], &idx, val)) {
+            uint64_t total = 0;
+            int i;
+            int end = (running_on_kernel < NETDATA_KERNEL_V4_15) ? 1 : ebpf_nprocs;
+            for (i = 0; i < end; i++)
+                total += val[i];
+
+            res[idx] = total;
+        } else {
+            res[idx] = 0;
+        }
+    }
+
+    socket_aggregated_data[0].call = res[NETDATA_KEY_CALLS_TCP_SENDMSG];
```

Comment:
We seem to be dragging this problem with fixed indexes around since beginning. Can we do something to fix this once for all? Use hashmap instead of array to store data etc. (so the elements will be named)? I think if we just use some magic values it will bite us in the future as the eBPF functionality is gradualy expanded.

Reply:
All right, I agree that this is good idea, but can I do this in another PR? And we concentrate effort here to separate the programs.

Reply:
OK I agree, we can do separate PR that cleans things up and organizes. Best would even be shared HDR for kernel side too. I will resolve this comment then.

Reply:
I created the issue #9204 for I bring this in a near future.